File: core/engine/excel_reader.py
import openpyxl

def read_test_data(file_path):
    """
    Read test data from an Excel file containing sheets:
    TEST_CASE, PAGE, ELEMENT, DATA
    Returns a dictionary with all the parsed data.
    """
    data = {
        "elements": {},
        "pages": {},
        "scenarios": [],
        "test_cases": [],
        "test_data": {}
    }
    
    try:
        workbook = openpyxl.load_workbook(file_path, data_only=True)
        
        # 1. Read ELEMENT
        if "ELEMENT" in workbook.sheetnames:
            sheet = workbook["ELEMENT"]
            headers = [cell.value for cell in sheet[1]]
            for row in sheet.iter_rows(min_row=2, values_only=True):
                if not row[0]: continue
                row_data = dict(zip(headers, row))
                element_key = row_data.get("ElementKey") or row_data.get("element_id")
                locator_type = row_data.get("LocatorType") or row_data.get("locator_type")
                locator_value = row_data.get("LocatorValue") or row_data.get("locator_value")
                if element_key:
                    data["elements"][element_key] = {
                        "locator": locator_value,
                        "locator_type": locator_type
                    }

        # 2. Read PAGE
        if "PAGE" in workbook.sheetnames:
            sheet = workbook["PAGE"]
            headers = [cell.value for cell in sheet[1]]
            for row in sheet.iter_rows(min_row=2, values_only=True):
                if not row[0]: continue
                row_data = dict(zip(headers, row))
                page_key = row_data.get("PageKey") or row_data.get("page")
                url = row_data.get("URL") or row_data.get("url")
                page_name = row_data.get("Page Name") or page_key
                if page_key:
                    data["pages"][page_key] = {"url": url, "name": page_name}

        # 3. Read DATA (support DATA and DATA_ prefix sheets)
        # We need to map datasets index-by-index per test_case_type
        # dataset indices tracks how many rows we have added for a given test_case_type
        dataset_indices = {}
        for sheet_name in workbook.sheetnames:
            if sheet_name == "DATA" or sheet_name.startswith("DATA_"):
                sheet = workbook[sheet_name]
                sheet_key = sheet_name.lower()
                headers = [cell.value for cell in sheet[1]]
                
                # Reset counts for each sheet so we match row 1 of DATA_A with row 1 of DATA_B
                sheet_dataset_indices = {}
                
                for row in sheet.iter_rows(min_row=2, values_only=True):
                    if not row[0]: continue # DataSet is empty
                    row_data = dict(zip(headers, row))
                    dataset = row_data.get("DataType") or row_data.get("DataSet") or row_data.get("test_case_type")
                    if dataset:
                        if dataset not in data["test_data"]:
                            data["test_data"][dataset] = []
                            
                        if dataset not in sheet_dataset_indices:
                            sheet_dataset_indices[dataset] = 0
                            
                        idx = sheet_dataset_indices[dataset]
                        
                        if idx >= len(data["test_data"][dataset]):
                            data["test_data"][dataset].append({"_flat": {}})
                            
                        # Store in sheet-specific namespace
                        data["test_data"][dataset][idx][sheet_key] = row_data
                        # Store flat for backward compatibility
                        data["test_data"][dataset][idx]["_flat"].update(row_data)
                        
                        sheet_dataset_indices[dataset] += 1

        # 4. Read TEST_CASE
        grouped_scenarios = {}
        current_tc_id = None
        current_dataset = None
        
        if "TEST_CASE" in workbook.sheetnames:
            sheet = workbook["TEST_CASE"]
            headers = [cell.value for cell in sheet[1]]
            for row in sheet.iter_rows(min_row=2, values_only=True):
                # if row is completely empty, skip
                if not any(row):
                    continue
                    
                row_data = dict(zip(headers, row))
                
                # Check if it's a new TC
                tc_id = row_data.get("TC_ID") or row_data.get("tc-id")
                if tc_id and str(tc_id).strip():
                    current_tc_id = str(tc_id).strip()
                    dataset_val = row_data.get("Data") or row_data.get("DataType") or row_data.get("DataSet") or row_data.get("type")
                    if dataset_val and str(dataset_val).strip():
                        current_dataset = str(dataset_val).strip()
                    else:
                        current_dataset = None
                    
                    if current_tc_id not in grouped_scenarios:
                        grouped_scenarios[current_tc_id] = {
                            "tc_id": current_tc_id,
                            "summary": row_data.get("Summary") or row_data.get("summary") or "",
                            "run": True,
                            "dataset": current_dataset,
                            "steps": []
                        }
                        
                if current_tc_id:
                    # check if this row has a step
                    step_no = row_data.get("Step") or row_data.get("step")
                    if step_no and str(step_no).strip():
                        # Add step to the current TC
                        # Notice we map headers from the image to standard internal keys
                        step_data = {
                            "TestCaseID": current_tc_id,
                            "StepNo": step_no,
                            "Action": row_data.get("Action") or row_data.get("action"),
                            "TargetElement": row_data.get("Target") or row_data.get("target"),
                            "DataColumn": row_data.get("Data") or row_data.get("value"),
                            "Expected": row_data.get("Expected") or row_data.get("expected")
                        }
                        grouped_scenarios[current_tc_id]["steps"].append(step_data)
                        
        data["test_cases"] = list(grouped_scenarios.values())

        return data
    except Exception as e:
        print(f"[Error] Failed to read Excel file '{file_path}': {e}")
        return data

File: framework/core/engine/excel/excel.reader.ts
import * as ExcelJS from 'exceljs';
import { FrameworkConfig } from '../../../config/framework.config';

export class ExcelReader {
    private static getCleanStringValue(val: any): any {
        if (val && typeof val === 'object') {
            if ('text' in val) val = (val as any).text;
            else if ('richText' in val) val = (val as any).richText.map((rt: any) => rt.text).join('');
            else if ('result' in val) val = (val as any).result; // for formulas
        }
        if (typeof val === 'string') {
            return val.replace(/_x000d_/g, '').trim();
        }
        return val?.toString() || val;
    }

    public static async readTestData(filePath: string): Promise<any> {
        const data: any = {
            elements: {},
            pages: {},
            scenarios: [],
            test_cases: [],
            test_data: {}
        };

        try {
            const workbook = new ExcelJS.Workbook();
            await workbook.xlsx.readFile(filePath);

            // 1. Read ELEMENT (ELEMENT and ELEMENT_ prefix)
            workbook.eachSheet((sheet) => {
                const sheetName = sheet.name.toUpperCase();
                if (sheetName === 'ELEMENT' || sheetName.startsWith('ELEMENT_')) {
                    const headers = sheet.getRow(1).values as string[];
                    sheet.eachRow((row, rowNumber) => {
                        if (rowNumber === 1) return;
                        const rowData: any = {};
                        for (let colNumber = 1; colNumber < headers.length; colNumber++) {
                            const cellValue = row.getCell(colNumber).value;
                            rowData[headers[colNumber]] = ExcelReader.getCleanStringValue(cellValue);
                        }
                        const elementKey = rowData['ElementKey'] || rowData['element_id'];
                        const locatorType = rowData['LocatorType'] || rowData['locator_type'];
                        const locatorValue = rowData['LocatorValue'] || rowData['locator_value'];
                        if (elementKey) {
                            data.elements[elementKey] = {
                                locator: locatorValue,
                                locator_type: locatorType
                            };
                        }
                    });
                }
            });

            // 2. Read PAGE
            const pageSheet = workbook.getWorksheet('PAGE');
            if (pageSheet) {
                const headers = pageSheet.getRow(1).values as string[];
                pageSheet.eachRow((row, rowNumber) => {
                    if (rowNumber === 1) return;
                    const rowData: any = {};
                    for (let colNumber = 1; colNumber < headers.length; colNumber++) {
                        const cellValue = row.getCell(colNumber).value;
                        rowData[headers[colNumber]] = ExcelReader.getCleanStringValue(cellValue);
                    }
                    const pageKey = rowData['PageKey'] || rowData['page'];
                    const url = rowData['URL'] || rowData['url'];
                    const pageName = rowData['Page Name'] || pageKey;
                    if (pageKey) {
                        data.pages[pageKey] = { url: url, name: pageName };
                    }
                });
            }

            // 3. Read DATA (DATA and DATA_ prefix)
            workbook.eachSheet((sheet, id) => {
                const sheetName = sheet.name;
                if (sheetName === 'DATA' || sheetName.startsWith('DATA_')) {
                    const sheetKey = sheetName.toLowerCase();
                    const headers = sheet.getRow(1).values as string[];
                    const localIndices: any = {};

                    sheet.eachRow((row, rowNumber) => {
                        if (rowNumber === 1) return;
                        
                        // Check if row has any non-empty cell
                        let isEmpty = true;
                        for (let colNumber = 1; colNumber < headers.length; colNumber++) {
                            const val = row.getCell(colNumber).value;
                            if (val !== null && val !== undefined && val !== '') {
                                isEmpty = false;
                                break;
                            }
                        }
                        if (isEmpty) return;

                        const rowData: any = {};
                        for (let colNumber = 1; colNumber < headers.length; colNumber++) {
                            const cellValue = row.getCell(colNumber).value;
                            rowData[headers[colNumber]] = ExcelReader.getCleanStringValue(cellValue);
                        }

                        const dataset = rowData['DataType'] || rowData['DataSet'] || rowData['test_case_type'];
                        if (dataset) {
                            if (!data.test_data[dataset]) {
                                data.test_data[dataset] = [];
                            }
                            if (localIndices[dataset] === undefined) {
                                localIndices[dataset] = 0;
                            }
                            const idx = localIndices[dataset];
                            if (idx >= data.test_data[dataset].length) {
                                data.test_data[dataset].push({ "_flat": {} });
                            }

                            data.test_data[dataset][idx][sheetKey] = rowData;
                            Object.assign(data.test_data[dataset][idx]["_flat"], rowData);

                            localIndices[dataset]++;
                        }
                    });
                }
            });

            // 4. Read all TEST_CASE sheets
            const groupedScenarios: any = {};

            workbook.eachSheet((tcSheet, id) => {
                if (!tcSheet.name.toUpperCase().startsWith('TEST_CASE')) {
                    return;
                }

                let currentTcId: string | null = null;
                let currentDataset: string | null = null;
                const headers = tcSheet.getRow(1).values as string[];

                tcSheet.eachRow((row, rowNumber) => {
                    if (rowNumber === 1) return;

                    const rowData: any = {};
                    let isEmpty = true;
                    for (let colNumber = 1; colNumber < headers.length; colNumber++) {
                        const cellValue = row.getCell(colNumber).value;
                        if (cellValue !== null && cellValue !== undefined && cellValue !== '') {
                            isEmpty = false;
                        }
                        rowData[headers[colNumber]] = ExcelReader.getCleanStringValue(cellValue);
                    }

                    if (isEmpty) return;

                    const tcId = rowData['TC_ID'] || rowData['tc-id'];
                    if (tcId && String(tcId).trim()) {
                        currentTcId = String(tcId).trim();
                        const datasetVal = rowData['Data'] || rowData['DataType'] || rowData['DataSet'] || rowData['type'];
                        currentDataset = (datasetVal && String(datasetVal).trim()) ? String(datasetVal).trim() : null;

                        const toRunVal = rowData['to_run'] || rowData['to-run'] || rowData['Run'] || rowData['run'];
                        const toRun = (toRunVal === undefined || toRunVal === null || String(toRunVal).trim() === '') ? false : (String(toRunVal).trim().toUpperCase() === 'Y');

                        const parameterizedVal = rowData['parameterized'] || rowData['parameterize'] || rowData['loop_from_step_1'] || rowData['loop'];
                        const isParameterized = (parameterizedVal && String(parameterizedVal).trim().toUpperCase() === 'Y');

                        if (!groupedScenarios[currentTcId]) {
                            groupedScenarios[currentTcId] = {
                                tc_id: currentTcId,
                                summary: rowData['Summary'] || rowData['summary'] || "",
                                run: toRun,
                                parameterized: isParameterized,
                                dataset: currentDataset,
                                sheet_name: tcSheet.name,
                                steps: []
                            };
                        }
                    }

                    if (currentTcId) {
                        const stepNo = rowData['Step'] || rowData['step'];
                        if (stepNo && String(stepNo).trim()) {
                            groupedScenarios[currentTcId].steps.push({
                                TestCaseID: currentTcId,
                                StepNo: stepNo,
                                Action: rowData['Action'] || rowData['action'],
                                TargetElement: rowData['Target'] || rowData['target'],
                                DataColumn: rowData['Data'] || rowData['value'],
                                Expected: rowData['Expected'] || rowData['expected']
                            });
                        }
                    }
                });
            });

            // Resolve custom validations lookup datasets from configuration dynamically
            data.custom_lookups = {};
            const customRules = FrameworkConfig.CUSTOM_VALIDATIONS || {};
            for (const [targetKey, rule] of Object.entries(customRules)) {
                if (rule && (rule as any).type === 'lookup') {
                    const sourceSheetName = (rule as any).sourceSheet;
                    const sourceColName = (rule as any).sourceColumn;
                    const values: string[] = [];

                    workbook.eachSheet((sheet) => {
                        const sheetNameUpper = sheet.name.toUpperCase();
                        const sourceSheetUpper = sourceSheetName.toUpperCase();
                        const isMatch = sheetNameUpper === sourceSheetUpper ||
                                        sheetNameUpper.startsWith(sourceSheetUpper + '_') ||
                                        sheetNameUpper.endsWith('_' + sourceSheetUpper) ||
                                        (sourceSheetUpper === 'DATA' && sheetNameUpper.startsWith('DATA'));

                        if (isMatch) {
                            const headers = sheet.getRow(1).values as string[];
                            let colIndex = -1;
                            for (let i = 1; i < headers.length; i++) {
                                if (headers[i] === sourceColName) {
                                    colIndex = i;
                                    break;
                                }
                            }
                            if (colIndex !== -1) {
                                sheet.eachRow((row, rowNumber) => {
                                    if (rowNumber === 1) return;
                                    const cell = row.getCell(colIndex);
                                    const strVal = ExcelReader.getCleanStringValue(cell.value);
                                    if (strVal && typeof strVal === 'string') {
                                        values.push(strVal);
                                    }
                                });
                            }
                        }
                    });
                    
                    data.custom_lookups[targetKey] = Array.from(new Set(values));
                }
            }
            data.valid_hospitals = data.custom_lookups['btn_dynamic_select'] || [];

            // Propagate sheet data from the first row to subsequent rows of the same dataset if missing
            for (const dataset of Object.keys(data.test_data)) {
                const rows = data.test_data[dataset];
                if (rows.length > 1) {
                    const allSheetKeys = new Set<string>();
                    rows.forEach((r: any) => {
                        Object.keys(r).forEach(k => {
                            if (k !== '_flat') allSheetKeys.add(k);
                        });
                    });
                    for (const sheetKey of allSheetKeys) {
                        let lastValidVal: any = null;
                        for (let i = 0; i < rows.length; i++) {
                            if (rows[i][sheetKey]) {
                                lastValidVal = rows[i][sheetKey];
                            } else if (lastValidVal) {
                                rows[i][sheetKey] = lastValidVal;
                                Object.assign(rows[i]["_flat"], lastValidVal);
                            }
                        }
                    }
                }
            }

            data.test_cases = Object.values(groupedScenarios);
            return data;

        } catch (e) {
            console.error(`[Error] Failed to read Excel file '${filePath}':`, e);
            return data;
        }
    }
}
